### ISSUE TITLE
Decode spends 22% of each step outside the transformer graph: 17 scalar DtoH round-trips, a 404 KB logits readback, and 48 allocator calls per token

### ISSUE THREAD

**Issue by justinchuby** [justinchuby/onnx-genai#1593]
Follow-up to #1574. With the GEMV measurement corrected in #1585, the GEMVs are at 41% of peak and the remaining decode budget is mostly *not* in them. This is what the rest is.

## Method

`nsys profile --cuda-graph-trace=node` (mandatory — without it the captured transformer is invisible), 96 tokens, idle A100-SXM4-80GB, 63 steady-state steps, medians. Cross-check: the GEMV kernels sum to 17.73 ms/token here against the probe's independent 18.40 ms, so the kernel accounting is sound.

## Where a decode step goes

| | ms/step | |
|---|---|---|
| **step total** | **29.58** | |
| `cuGraphLaunch` host submit | 2.79 | |
| blocking sync on the graph | 22.85 | GPU running the transformer |
| chatty per-op CUDA tail | 3.76 | 19 DtoH, 48 alloc/free, 19 eager launches |
| host-only dead time | 0.13 | |

**6.55 ms/step — 22% — is outside the transformer.** The graph itself is dense: kernels sum to 22.58 ms against a 22.85 ms sync, so only 0.3 ms of intra-graph gap. There is nothing to win inside the capture.

## The chatty tail, in detail

Per step, 19 DtoH copies totalling 417 KB:

| size | count | what it must be |
|---|---|---|
| 8 B | 14 | scalar readback |
| 1 B | 2 | scalar readback |
| 4 B | 1 | scalar readback |
| 13312 B | 1 | one bf16 hidden state (6656 x 2) |
| 404096 B | 1 | **the full logits vector** (202048 x 2) |

**17 of the 19 copies move eight bytes or fewer**, and each is preceded by its own `cuStreamSynchronize`. That is 17 full host-to-GPU round-trips per token to read scalars. The eager kernels around them (`cumsum_i64`, `equal_i64`, `where_bytes`, `gather_bytes`, `clip_i64`, `or_bool`, `add_i64`, `sub_i64`, `reduce_i64`) are the position-id/attention-mask bookkeeping and the embedding gather — the classic `position_ids = cumsum(mask) - 1` shape, executed op by op on the host with a drain after each.

Also per step: **48 `cuMemAlloc`/`cuMemFree` calls** through the VMM arena, for intermediates that are the same size every token.

## Why this is worth fixing

The 3.76 ms tail is 13% of the step and none of it is arithmetic. Unlike the GEMVs — which are at 41% of a hard bandwidth ceiling and have maybe 1.5x left — this is pure serialization, and the work it protects is a handful of integer ops on tensors of a few elements.

Ordered by expected value:

1. **Stop reading scalars back.** The 17 scalar DtoH copies exist to drive host-side control flow (dynamic shapes, sequence bookkeeping). Decode shapes are static per step; these values should stay resident and be consumed by device-side kernels.
2. **Sample on the device.** The 404 KB logits readback drains the pipeline every token to run argmax/sampling on the host. A device-side sampler returns 4 bytes instead of 404 KB and removes the drain. Note the related win already identified in #1574: lm_head only needs the last position's logits.
3. **Capture the tail, or pool its buffers.** 48 allocator calls per token for fixed-size intermediates is churn the arena should absorb; and once the scalar readbacks are gone, the whole tail becomes capturable.
4. **`cuGraphLaunch` costing 2.79 ms of host time** needs its own look — a graph submit is normally tens of microseconds. Some of this is nsys node-tracing overhead, so it should be confirmed with a clean timing before anyone acts on it.

## Reproduce

```bash
CUDA_VISIBLE_DEVICES=<idle gpu> ONNX_GENAI_CUDA_DEVICE=0 \
  /usr/local/cuda/bin/nsys profile --cuda-graph-trace=node --trace=cuda \
  -o /tmp/n96 ./target/release/profile_native \
  --model <model> --pipeline --ep cuda --backend native --tokens 96 --warmups 0 --runs 1
/usr/local/cuda/bin/nsys export --type sqlite -o /tmp/n96.sqlite /tmp/n96.nsys-rep
```

Then delimit steps by `cuGraphLaunch` in `CUPTI_ACTIVITY_KIND_RUNTIME` and take medians over steady-state steps. See `.github/skills/cuda-perf-measurement` before trusting any number from this.

---

**Update (#1592):** the 404096 B per-token logits readback listed above is now
gone for greedy decoding. Two gates were closing the device argmax on this path:
`PipelineDecodeLoopBackend` never overrode `greedy_fastpath_supported()`, and the
eligibility test was `chain.is_empty()` when every real model ships `top_k` /
`top_p` defaults plus a chat-template stop sequence. Both are fixed; the test is
now argmax-equivalence. Worth ~1-2% here (the readback was ~0.6 ms of a ~30 ms
step), and `loop.sampling` disappears entirely.

Still open from the list above: the 17 scalar DtoH round-trips, the 48
allocator calls per token, and pulling the eager tail into the capture.
